### PR TITLE
Add ability to perform variable substitution to stack links in shipit.yml

### PR DESCRIPTION
**Pull request by indiebrain**
...when at least one task has been performed on the stack - IE when
we have something we can use to grab the environment variables.

This maintains the previous functionality of the `shipit.yml` to
declare static links for stacks like:

```yml
links:
  logs: https://logs.application.com
```

but also affords consumers the ability to define template strings in
the `shipit.yaml` like:

```yml
links:
  logs: https://logs.$ENVIRONMENT.application.com
```

Which, if the stack's environment is set to `production` will render a
link to `https://logs.production.application.com` in the stack's header.

### CODE REVIEW

File: app/models/shipit/stack.rb [hunk 1]
```diff
@@ -528,8 +528,25 @@ def sync_github
       GithubSyncJob.perform_later(stack_id: id)
     end
 
+    alias links_orig links
```
Comment (casperisfine) [inline app/models/shipit/stack.rb]: Rather than aliasing the method, you can remove it from the delegation list, and call `cached_deploy_spec.links`.

Also note that this `delegate` has a `allow_nil: true`, so you have to handle `cached_deploy_spec` being nil.

File: app/models/shipit/stack.rb [hunk 2]
```diff
@@ -528,8 +528,25 @@ def sync_github
...
+    def links
+      persisted_tasks = tasks.select(&:persisted?)
+      return links_orig if persisted_tasks.blank?
+
+      variables = Shipit::TaskCommands.new(persisted_tasks.last).env
+      links_orig.each_with_object({}) do |(name, url), replaced_links|
```
Comment (casperisfine) [inline app/models/shipit/stack.rb]: `Hash#transform_values` would be much cleaner here.

File: app/models/shipit/stack.rb [hunk 3]
```diff
@@ -528,8 +528,25 @@ def sync_github
       GithubSyncJob.perform_later(stack_id: id)
     end
 
+    alias links_orig links
+    def links
+      persisted_tasks = tasks.select(&:persisted?)
+      return links_orig if persisted_tasks.blank?
+
+      variables = Shipit::TaskCommands.new(persisted_tasks.last).env
```
Comment (casperisfine) [inline app/models/shipit/stack.rb]: I have to think about that one. This seems quite a hack.

What variable do you need to interpolate here? 

I'm not against that interpolation, but if it could be done without loading the last deploy it would be preferable.

File: app/models/shipit/stack.rb [hunk 4]
```diff
@@ -528,8 +528,25 @@ def sync_github
...
+      links_orig.each_with_object({}) do |(name, url), replaced_links|
+        replaced_links[name] = replace_variables(url, variables)
+      end
+    end
+
     private
 
+    def replace_variables(string, variables)
```
Comment (casperisfine) [inline app/models/shipit/stack.rb]: We already have code for that: https://github.com/Shopify/shipit-engine/blob/4192f63825a6e7cf8f437cb2fa0f626ad60f4cf2/lib/shipit/command.rb#L48-L55

Might just need to be extracted to it's own module.

File: app/models/shipit/stack.rb (hunk 3 above)
Comment (indiebrain) [reply app/models/shipit/stack.rb]: Perhaps a practical example will be helpful. 

We run and deploy to a number of internal Kubernetes clusters. We have a custom task which chooses the _least active_ cluster and provisions a Kubernetes namespace for a stack when a new stack is added. We register internal hooks to deployments which capture information about the kubernetes cluster to which a stack is deployed and make that knowledge available via the deployment/task's `#env`. 

This information is useful to us because the cluster to which a stack is deployed determines the URL of its telemetry services - for example the pattern to get to a stack's logs looks something like `https://logs.${ENVIRONMENT}.application.${CLUSTER}.domain.com`. Does that help at least put a frame around the picture?

File: app/models/shipit/stack.rb (hunk 2 above)
Comment (indiebrain) [reply app/models/shipit/stack.rb]: Wow, `Hash#transform_values` is much nicer. Thanks for the suggestion!

File: app/models/shipit/stack.rb (hunk 1 above)
Comment (indiebrain) [reply app/models/shipit/stack.rb]: Addressed by https://github.com/Shopify/shipit-engine/pull/1018/commits/c2ec62ac53841a7e3d2f32e1530916a94428e964

File: app/models/shipit/stack.rb (hunk 2 above)
Comment (indiebrain) [reply app/models/shipit/stack.rb]: Addressed by https://github.com/Shopify/shipit-engine/pull/1018/commits/ae342cc1e1b9e5447736efa569d88aa0b67e6ead

File: app/models/shipit/stack.rb (hunk 4 above)
Comment (indiebrain) [reply app/models/shipit/stack.rb]: Addressed by https://github.com/Shopify/shipit-engine/pull/1018/commits/671d045de9eb4d64cd2e4dcc9e7bb8a9ac5cdf97

File: app/models/shipit/stack.rb (hunk 3 above)
Comment (casperisfine) [reply app/models/shipit/stack.rb]: > We register internal hooks to deployments which capture information about the kubernetes cluster to which a stack is deployed and make that knowledge available via the deployment/task's `#env`.

Ok, so you use tasks to persist data?

It's quite a clever hack, but from a product design standpoint, using the last stack for interpolation would make for a very surprising behavior.

What if:

  - We use `Stack#env` for interpolation (doesn't exist yet, but easy to add).
  - You can then do this cluster selection in your host app in ruby, and store it either in the db or in `Rails.cache`.

Comment (indiebrain) [reply app/models/shipit/stack.rb]: > We use Stack#env for interpolation (doesn't exist yet, but easy to add).

Sounds like something I can chase down. Just want to make sure I'm on the same page. If this bit lives in `shipit-engine` what would your recommendation be to allow the host app - like mine - inject values into the stack's environment? Would it be appropriate to add another concept / association for these values to the Stack and use them to build the value of `Stack#env`? Perhaps setting one of these values in the host application looks something like `stack.envrionment_variable.create(name: "cluster", value: "cluster-1")`? Does this seem reasonable? Do you think the name `environment_variable` is appropriate here? If not, any suggestions?

Comment (casperisfine) [reply app/models/shipit/stack.rb]: IMO the best way is:

```ruby
module ExtraStackEnv
  def env
    env = super
    # add whatever you want here
    env
  end
end

Shipit::Stack.prepend(ExtraStackEnv)
```

Comment (indiebrain) [reply app/models/shipit/stack.rb]: Addressed by https://github.com/Shopify/shipit-engine/pull/1018/commits/bc0fd9d791729cfc0eb497c764f53df53112b38e and https://github.com/Shopify/shipit-engine/pull/1018/commits/a210081966d860400107e216f80446ab2f1f366a

**Comment by indiebrain** [comment]
@casperisfine something like this agreeable?

**Comment by casperisfine** [comment]
Yes. SOrry I started reviewing this hours ago and somehow got distracted. Reviwing it gain .

**Comment by indiebrain** [comment]
👍 thanks!

File: lib/shipit/task_commands.rb [hunk 5]
```diff
@@ -28,22 +28,20 @@ def steps
...
+          'BUNDLE_PATH' => Rails.root.join('data', 'bundler').to_s,
+          'SHIPIT_LINK' => @task.permalink,
+          'TASK_ID' => @task.id.to_s,
+          'IGNORED_SAFETIES' => @task.ignored_safeties? ? '1' : '0',
+          'GIT_COMMITTER_NAME' => @task.user&.name || Shipit.committer_name,
+          'GIT_COMMITTER_EMAIL' => @task.user&.email || Shipit.committer_email,
+        )
+        .merge(deploy_spec.machine_env)
+        .merge(@stack.env)
+        .merge(@task.env)
```
Comment (casperisfine) [inline lib/shipit/task_commands.rb]: My last worry with the PR is here.

Environment variables precedence is important.

I don't think `Stack#env` should be able to override `Task#env`.

Intuitively I'd say the order should be:

  - `Commands#env` (`super` in this context, mostly comes from the global config file)
  - `Stack#env`
  - `TaskCommands#env`
  - `DeploySpec#machine_env`
  - `Task#env`

But I'm open to diverging opinions.

Comment (indiebrain) [reply lib/shipit/task_commands.rb]: Entirely reasonable. Thanks for the feedback! Something like https://github.com/Shopify/shipit-engine/pull/1018/commits/09cc2991332935537453101c112f6855faf72a84 instead, then?

Comment (casperisfine) [reply lib/shipit/task_commands.rb]: Yup 👍

Comment (indiebrain) [reply lib/shipit/task_commands.rb]: Many, many thanks, friend. 🍻

File: test/dummy/db/schema.rb [hunk 6]
```diff
@@ -10,7 +10,7 @@
 #
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema.define(version: 2020_01_09_132519) do
+ActiveRecord::Schema.define(version: 2020_02_11_145825) do
```
Comment (indiebrain) [inline test/dummy/db/schema.rb]: Whoops. @casperisfine . Not sure why these changes weren't removed in https://github.com/powerhome/shipit-engine/commit/a210081966d860400107e216f80446ab2f1f366a

Comment (casperisfine) [reply test/dummy/db/schema.rb]: I should have seen it. No big deal.

Comment (indiebrain) [reply test/dummy/db/schema.rb]: My apologies.

Comment (casperisfine) [reply test/dummy/db/schema.rb]: No worries. I'll fix that on master.

Comment (indiebrain) [reply test/dummy/db/schema.rb]: You're a gentle-person and a scholar.
